### PR TITLE
Template editor React implementation

### PR DESCRIPTION
**Pull request by kienstra**
Updated April 22nd

#### Changes

* This syntax in the template: `{{cta-text}}` would be: `block_field( 'cta-text' )`
* But that's the only dynamic syntax. This doesn't have conditionals or for loops, it's not even Mustache
* There's a Markup and CSS editor:
markup-button-here

#### Testing Instructions

1. `composer install && npm install && npm run dev`
2. `/wp-admin` > Custom Blocks > Add New
3. Give the block any title you want: 
your-blocks-
4. Add a new field, maybe named "Your Example Field": 
your-example-field
5. Click the 'Template Editor' button
6. Expected: It should reference the field you created: 
your-example-one
7. Enter any markup you want, maybe like:

```html
<div class="foo-class">
    <p class="baz">
        {{your-example-field}}
    </p>
</div>
```
some-markup

8. Click 'Save draft' and refresh the page
9. Expected: The markup should still be there
10. Click the 'CSS' button and enter any CSS
11. Click 'Save draft'
12. Expected: it's still there

Out of scope: 'Front-end Preview' will not work for this. That's in an upcoming PR.  This doesn't do anything with the markup or CSS editor yet, it just saves it.

Fixes [GF-2974](https://wpengine.atlassian.net/browse/GF-2974)

### CODE REVIEW

**Comment by kienstra** [comment]
With Phil's ideas, I'll probably drastically change this.

File: js/src/edit-block/components/clipboard-copy.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 /**
  * External dependencies
  */
-import React from 'react';
+import * as React from 'react';
```
Comment (kienstra) [inline js/src/edit-block/components/clipboard-copy.js:4]: To match how React is imported in the rest of this plugin

File: js/src/edit-block/components/header.js [hunk 2]
```diff
@@ -88,6 +89,17 @@ const Header = ( { editorMode, setEditorMode } ) => {
...
+			<button
+				className={ classNames(
+					buttonClasses,
+					{ 'font-semibold': TEMPLATE_EDITOR_EDITING_MODE === editorMode }
+				) }
+				onClick={ () => {
+					setEditorMode( TEMPLATE_EDITOR_EDITING_MODE );
+				} }
+			>
+				<span>{ __( 'Template Editor', 'genesis-custom-blocks' ) }</span>
```
Comment (kienstra) [inline js/src/edit-block/components/header.js]: Maybe 'Template' isn't the right word

File: package.json [hunk 3]
```diff
@@ -29,34 +29,35 @@
...
+    "@wordpress/hooks": "2.12.0",
+    "@wordpress/html-entities": "2.11.0",
+    "@wordpress/i18n": "3.19.0",
+    "@wordpress/icons": "2.10.0",
+    "@wordpress/keycodes": "2.19.0",
+    "@wordpress/notices": "2.13.0",
+    "@wordpress/postcss-plugins-preset": "2.1.0",
+    "@wordpress/scripts": "14.0.0",
+    "@wordpress/server-side-render": "1.19.5",
+    "@wordpress/url": "2.22.0",
```
Comment (kienstra) [inline package.json:60]: These updates were needed to get Jest to pass

File: tests/js/jest.config.js [hunk 4]
```diff
@@ -4,7 +4,6 @@ module.exports = {
 	transform: {
 		'^.+\\.[jt]sx?$': '<rootDir>/node_modules/@wordpress/scripts/config/babel-transform',
 	},
-	testEnvironment: 'jest-environment-jsdom-sixteen',
```
Comment (kienstra) [inline tests/js/jest.config.js:7]: It looks like this isn't needed

**Comment by kienstra** [comment]
Todo: Look at why the latest [@wordpress/server-side-render](https://npmjs.com/package/@wordpress/server-side-render) isn't working with this. 

That could be a big problem, we might even need to fork that if that package version makes it into Core.

Update 2 April: The latest (`@wordpress/server-side-render@1.21.0`) now seems to work fine. I'm not sure what happened.

**Comment by kienstra** [comment]
@johnstonphilip,
If you'd like, feel free to branch off this. 

So far, this PR is mainly to get CircleCI to pass, with just a little skeleton code for the feature.

**Comment by kienstra** [comment]
The template editor is partly working now, though it needs an `onChange` handler to save the template:

the-template-editor

The PHP rendering could parse this, and whenever it finds `{{foo}}`, it could do `block_field( 'foo' )`. 

Otherwise, it could just render the plain HTML.

**Comment by kienstra** [comment]
So this in the template editor:

```html
<div>
    <h2>{{hero-heading}}</h2>
    <div>
        <img src="{{hero-image}}" alt="{{hero-image-alt}}" />
        <div>
            <p>
                {{hero-copy}}
            </p>
        </div>
        <div>
            <a href="{{cta-url}}">
                {{cta-text}}
            </a>
        </div>
    </div>
</div>
```

Would be rendered as this in PHP:

```php
<div>
    <h2><?php block_field( 'hero-heading' ); ?></h2>
    <div>
        <img src="<?php block_field( 'hero-image' ); ?>" alt="<?php block_field( 'hero-image-alt' ); ?>" />
        <div>
            <p>
                <?php block_field( 'hero-copy' ); ?>
            </p>
        </div>
        <div>
            <a href="<?php block_field( 'cta-url' ); ?>">
                <?php block_field( 'cta-text' ); ?>
            </a>
        </div>
    </div>
</div>
```

I'd suggest we don't even use [Mustache](https://mustache.github.io/#demo). We could just say that `{{foo}}` will echo the value of the `'foo'` field.

**Comment by kienstra** [comment]
We'd also need handling for when to use the existing PHP templates.

For example, if there's markup in the React editor, it wouldn't render the PHP template, nor show a notice for it if there isn't one:

here-is-the-template-notice

**Comment by kienstra** [comment]
This work could have 3 parts:

1. React, which this PR starts
2. [Front-end PHP rendering](https://github.com/studiopress/genesis-custom-blocks/pull/65#issuecomment-811618913), not in this PR at all

3. Testing them together with an e2e test and manual exploratory testing

Steps 1 and 2 could be done in parallel by different developers.

There will probably be more UI touches needed, like explaining how `{{}}` works, and supporting the existing PHP templates if the GUI template editor is empty.

We'll also have to let Susan know so she can update the documentation. I think I gave her 7-10 working days' notice last time.

This will probably be the biggest doc change this plugin has had.

**Comment by kienstra** [comment]
Hm, I'm not sure why the e2e tests are failing. They're passing locally.

**Comment by kienstra** [comment]
The e2e tests started failing with https://github.com/studiopress/genesis-custom-blocks/pull/65/commits/451fe34fab04c93bbd3b77245538f0cca4ea08c3, and they pass when it's reverted.

File: package.json [hunk 5]
```diff
@@ -106,8 +108,8 @@
   },
   "scripts": {
     "check-licenses": "wp-scripts check-licenses --prod --gpl2",
-    "build": "wp-scripts build",
-    "dev": "wp-scripts start",
+    "build": "wp-scripts build --webpack-no-externals",
+    "dev": "wp-scripts start --webpack-no-externals",
```
Comment (kienstra) [inline package.json:112]: To ensure this uses the npm package versions in `package.json`, not whatever is in `wp.blocks`, `wp.components`, etc...

**Comment by Alynva** [comment]
Would be nice to have the styling editor too. Awesome work

**Comment by kienstra** [comment]
Hi @Alynva,
Thanks!!! Interesting, I'll keep that in mind about the styling editor.

**Comment by kienstra** [comment]
## 3 possible options for the CSS editor


1. Tabs for 'Markup' and 'CSS.' Clicking one reveals the markup or CSS editor:
here-are-the-ones

2. Markup editor above the CSS editor: 
the-markup-editor

3. A separate tab for the CSS editor at the top: 
bib-bisdf

**Comment by kienstra** [comment]
Hi @RobStino,
What do you think of the added syntax highlighting? It's still not a lot different, but maybe a bit more.

## Before

before-syntax-highlighting


## After
with-more-syntax-highlighting

**Comment by kienstra** [comment]
Here's how this looks so far:

## Markup Editor

markup-button-here

## CSS Editor

the-scc

**Comment by kienstra** [comment]
Latest screenshots:

# Markup

the-markup-selected

# CSS

css-button

File: .stylelintrc.json [hunk 6]
```diff
@@ -1,5 +1,5 @@
 {
-	"extends": "stylelint-config-wordpress",
+	"extends": "@wordpress/stylelint-config",
```
Comment (kienstra) [inline .stylelintrc.json:2]: A lot of edits here are to get linting and tests to pass now that packages are updated, like `@wordpress/scripts`

File: tests/e2e/specs/template-editor.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+		await page.keyboard.type( blockName );
+
+		const $editBlockDocument = await getDocument( page );
+		await ( await findByLabelText( $editBlockDocument, 'Add a new field' ) ).click();
+		await findByLabelText( $editBlockDocument, 'Field Label' );
+
+		await page.keyboard.type( field.label );
+		await page.select( '#field-control', field.type );
+		await ( await findByText( $editBlockDocument, 'Template Editor' ) ).click();
+		( await page.waitForSelector( '#gcb-template-editor' ) ).type( templateMarkup );
```
Comment (kienstra) [inline tests/e2e/specs/template-editor.js:36]: I'll add assertions here once the PHP rendering is added in the next sprint.

There isn't much to test in the template editor now, just that you can enter something into it.

**Comment by johnstonphilip** [comment]
@kienstra I'm not sure why yet, but I am getting these errors in the console:
Screen Shot 2021-04-23 at 10 12 54 AM

**Comment by kienstra** [comment]
@johnstonphilip,
Thanks, I see those console errors too. I'm looking at them.

**Comment by kienstra** [comment]
@johnstonphilip,
Thanks for pointing that out, do those commits help?

There don't look to be console errors in my local.

**Review by dreamwhisper — APPROVED** [review]
This is working well for me with no console errors.

A couple things for improvement in the future:
- Autosuggesting or otherwise showing what field names are available. 
- Improving the accessibility of buttons. Right now, if I'm tabbing through the buttons, there  is no indicator showing where the focus is. The GCB sidebar tabs also have no focus changes so could address that in a wider pass.

**Comment by kienstra** [comment]
Hi @dreamwhisper,
Thanks, great points! 

https://github.com/studiopress/genesis-custom-blocks/pull/65/commits/a0d2c280b606dd02e04ffc212496f85d2ef4d0ac removes the `:focus outline: none` styling. 

>Autosuggesting or otherwise showing what field names are available.

That's a good idea. I'm not sure of a great way to make it work in the editor, maybe showing the fields above the editor would be good.

**Comment by Alynva** [comment]
Sorry for the intrusion in this project, I'm really interested in how this is going.

> > Autosuggesting or otherwise showing what field names are available.
> 
> That's a good idea. I'm not sure of a great way to make it work in the editor, maybe showing the fields above the editor would be good.

Using the Ace Editor, you can just implement a new completer receiving the prefix and calling the callback with the suggestion list: https://github.com/ajaxorg/ace/wiki/How-to-enable-Autocomplete-in-the-Ace-editor . Here there is an example using the React component: https://github.com/securingsincity/react-ace/issues/69#issuecomment-354134011

I'm not sure where and how the field names are stored, but from the code editor perspective, I'm quite sure that's not a hard work.

**Comment by kienstra** [comment]
@Alynva,
Thanks a lot for that! I appreciate your help, no need to apologize 😄 

Thanks, your autocompleter example in React is great: https://github.com/securingsincity/react-ace/issues/69#issuecomment-354134011

It might not make it into this PR, I might have to think about it this week.

**Pull request merged**